Accessibility:  change table layout in toolbar

Darshan-upadhyay1110 opened the pull request:
Change table layout to CSS base table structure

- For the editor in the home view, a layout table is used, and it includes the use of IDs. This can
potentially lead to issues with the interpretation of content by assistive technologies
- Layout tables are
meant to structure the layout of a page and should not contain structural markup like th , caption ,
summary , headers , or id.

Solution:
- replace `td` and 'table' tag with `div`
- make some changes on css to be consistent with our previous design

Adjust table elements for mobile view
- we changed table structure of `toolbar-wrapper` with `div`
- so we also need to consider mobile view
- made some necessary changes for mobile view because of structure change in cool.html.m4 ('toolbar-wrapper'

eszkadev commented on browser/src/control/Control.MobileTopBar.js:
```diff
@@ -205,9 +205,7 @@ L.Control.MobileTopBar = L.Control.extend({
-				toolbar.hide('PermissionMode');
-				toolbar.hide('after-PermissionMode');
-				$('#tb_actionbar_item_before-PermissionMode').width('');
+				$('#MobilePermissionMode').hide();
```
note: please try to avoid using jQuery in the new code if possible.
For example you can use:
`document.querySelector('#MobilePermissionMode')`
element.style.display = 'none' for hide
element.style.display = '' for show

Darshan-upadhyay1110 replied:
Updated. Thanks @eszkadev 🙇🏼

eszkadev reviewed (commented):
code is ok, waiting for Pedro wrt CSS

eszkadev commented:
I canceled cypress, I will retrigger build after my fix for cypress will be merged.

eszkadev commented:
rebased so it will include toolbar reworks

pedropintosilva reviewed (changes requested):
Desktop: all good
Mobile:
* the back button and hamburger are misaligned -> fixed pushed another commit
* Calc on mobile is broken: formula bar is inaccessible: I have pushed (last commit [WIP]) that fixes almost all issues, please feel free to re-use it and fix the last cosmetic issue

pedropintosilva commented:
> Desktop: all good Mobile:
> 
>     * the back button and hamburger are misaligned -> fixed pushed another commit
> 
>     * Calc on mobile is broken: formula bar is inaccessible: I have pushed (last commit [WIP]) that fixes almost all issues, please feel free to re-use it and fix the last cosmetic issue

Tested on real android device with nc with chrome and chrome remote debug tools

Darshan-upadhyay1110 commented:
> Desktop: all good Mobile:
> 
> * the back button and hamburger are misaligned -> fixed pushed another commit
> * Calc on mobile is broken: formula bar is inaccessible: I have pushed (last commit [WIP]) that fixes almost all issues, please feel free to re-use it and fix the last cosmetic issue

- I think we should not change flex to grid, it is risky.
- I have tested this it moves the toolbar icons if we type in formula bar
Here is the ref:
[Screencast from 2024-03-29 16-19-32.webm](https://github.com/CollaboraOnline/online/assets/61383886/c59e7cf8-7b53-46a0-a790-cfbdf4a145f4)


I have made some changes by adding a wrapper with CSS.

I have tested all apps in desktop + mobile. All Good

CC: @pedropintosilva

eszkadev commented:
Please rebase on top of recent w2ui rework to be sure it still works.

Darshan-upadhyay1110 commented:
> Please rebase on top of recent w2ui rework to be sure it still works.

Things are now broken after w2ui rework :(. I will try to fix it

CC: @pedropintosilva please do not review it now i will post comment after fixes thanks

Darshan-upadhyay1110 commented:
> > Please rebase on top of recent w2ui rework to be sure it still works.
> 
> Things are now broken after w2ui rework :( . I will try to fix it
> 
> CC: @pedropintosilva please do not review it now i will post comment after fixes thanks

It seems a normal fix. There is fix height for toolbar-up, which is not good. Tested with mobile view and desktop all looks OK

CC: @pedropintosilva ready to review

eszkadev commented:
@Darshan-upadhyay1110 there are incoming mobile toolbar conversions, best to not touch mobile yet
After we will remove w2ui then we can do it on clean structure

Darshan-upadhyay1110 commented:
> @Darshan-upadhyay1110 there are incoming mobile toolbar conversions, best to not touch mobile yet After we will remove w2ui then we can do it on clean structure
> 
> eg PR: #8673

Sure, maybe i will wait till this PR merge then will proceed to change table thing

eszkadev commented on browser/src/control/Control.MobileTopBar.js:
```diff
@@ -122,18 +122,20 @@ L.Control.MobileTopBar = L.Control.extend({
 				this.enableItem(id, true);
 			});
-			this.showItem('PermissionMode', false);
+			PermissionMode.style.display = 'none';
```
`showItem` didn't work as expected? I prefer to use high level abstration. (if it doesn't work it's ok, it's HTML content we need to convert in future into some "native" widget)

Darshan-upadhyay1110 replied:
The problem is `PermissionMode` is not part of parent Container, so that is why i change showItem to simple html hide

Darshan-upadhyay1110 replied:
We can change here showItem method a bit to pass parent container and then can use to show or hide element using ShowItem. @eszkadev thoughts ?

Darshan-upadhyay1110 replied:
Something like this
![image](https://github.com/CollaboraOnline/online/assets/61383886/cba87d0b-a687-4318-9609-5571aa7bd48a)

here for element like permissionMode which now extracted outside (#Toolbar-up) can have it's own parent to pass on to showItem

eszkadev commented on browser/src/control/Control.MobileTopBar.js:
```diff
@@ -122,18 +122,20 @@ L.Control.MobileTopBar = L.Control.extend({
 	onUpdatePermission: function(e) {
 		var toolbarButtons = ['next', 'prev', 'mobile_wizard', 'insertion_mobile_wizard', 'comment_wizard'];
+		// Handle permission mode as diffrent case because it's not part of `toolbar-up` element
+		var PermissionMode = document.getElementById('PermissionMode');
```
variable name should start with small character

eszkadev commented on browser/html/cool.html.m4:
```diff
@@ -201,21 +201,22 @@ m4_ifelse(MOBILEAPP,[true],
+          <div id="toolbar-logo"></div>
+          <div id="toolbar-mobile-back" class="editmode-off"></div>
+          <div id="toolbar-mobile-center-space">
+            <div id="PermissionMode" class="cool-font status-readonly-mode" style="padding: 3px 10px; display: none"> Read-only</div>
```
please remove this `PermissionMode` thing, it should be only generated by JSDialogs builder

Darshan-upadhyay1110 commented:
Note: Upstream this [https://github.com/CollaboraOnline/online/pull/8688] before start to rework on current PR

Darshan-upadhyay1110 commented:
Note:

cypress_test/integration_tests/desktop/writer/scrolling_spec.js seems to be a strange test case. as a locally i tried the same document which is used in this document and followed the same steps.

I can see in status bar there is a possibility to get "Page 1 and 2 of 4" it can not be always a specific string like "Page 1 of 4" or "Page 2 of 4" it can "Page 2 and 3 of 4"

This is a real case that I tried.

eszkadev commented:
Please rebase on top of recent w2ui rework to be sure it still works.

> Note:
> 
> cypress_test/integration_tests/desktop/writer/scrolling_spec.js seems to be a strange test case. as a locally i tried the same document which is used in this document and followed the same steps.
> 
> I can see in status bar there is a possibility to get "Page 1 and 2 of 4" it can not be always a specific string like "Page 1 of 4" or "Page 2 of 4" it can "Page 2 and 3 of 4"
> 
> This is a real case that I tried.

For me it happened when I changed size of toolbars (height), then view was different and it reported different state fro "currently visible pages"

Darshan-upadhyay1110 commented:
> Please rebase on top of recent w2ui rework to be sure it still works.
> 
> > Note:
> > cypress_test/integration_tests/desktop/writer/scrolling_spec.js seems to be a strange test case. as a locally i tried the same document which is used in this document and followed the same steps.
> > I can see in status bar there is a possibility to get "Page 1 and 2 of 4" it can not be always a specific string like "Page 1 of 4" or "Page 2 of 4" it can "Page 2 and 3 of 4"
> > This is a real case that I tried.
> 
> For me it happened when I changed size of toolbars (height), then view was different and it reported different state fro "currently visible pages"
Rebased
No change, still that case is failed.

eszkadev commented:
Are you sure you didn't change any toolbar size? even 1 px, please check canvas size used in that test

eszkadev reviewed (changes requested):
one commit has name: ` WIP Please Darshan take this: Calc: Mobile: top toolbar and formula ` could you rename it or squash with other?
So it will not look like mistake.

Darshan-upadhyay1110 commented:
> 2 changes requested

Done

eszkadev reviewed (approved):
ok, then I will leave final decision for @pedropintosilva as it is mainly css change

Darshan-upadhyay1110 commented:
Tested mobile test case all are passing in local build. 
![image](https://github.com/CollaboraOnline/online/assets/61383886/206353bb-f264-41fc-ab5a-83601c696a3c)

pedropintosilva reviewed (approved):
thanks, there are still other things that need fix/improvement but lets get this merged and I will send another PR and ask your testing @Darshan-upadhyay1110

Darshan-upadhyay1110 commented on browser/html/cool.html.m4:
```diff
@@ -201,21 +201,20 @@ m4_ifelse(MOBILEAPP,[true],
+          <div id="toolbar-mobile-back" class="editmode-off"></div>
+          <div class="jsdialog ui-spacer"></div>
+          <div id="toolbar-up"></div>
+          <div id="toolbar-hamburger">
+            <label class="main-menu-btn" for="main-menu-state">
+              <span class="main-menu-btn-icon" id="main-menu-btn-icon"></span>
+            </label>
+          </div>
```
Sure
